### PR TITLE
Add service and state for saving very minimal app and unit records in dqlite

### PR DESCRIPTION
**Pull request by wallyworld**
This PR adds *very* minimal support for saving and deleting apps and units in dqlite. It does just enough to create the rows so that when storage is added, there's a unit which can act as the storage owner.

The next PR will wire this up so that when apps and units are written to mongo, we also call the services in this PR to write to dqlite as well.

The params struct used to hold the args to create a unit is very minimal - just the unit name. There are 2 of them - one to use for external callers when invoking the service method, the other for the service to use to pass args to state. They have the same content right now but will diverge as things get fleshed out.

To allow for adding multiple units and only incurring the cost of preparing the sqlair statements once, a closure is used to facilitate the insert of the unit rows in the same transaction.

## QA steps

Just unit tests.

## Links

**Jira card:** JUJU-5392

### CODE REVIEW

File: domain/application/errors/errors.go [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+package errors
+
+import (
+	"github.com/juju/errors"
+)
+
+const (
+	// NotFound describes an error that occurs when the application being operated on
+	// does not exist.
+	NotFound = errors.ConstError("application not found")
```
Comment (SimonRichardson) [inline domain/application/errors/errors.go:13]: I'm unsure we want this, why not just use domain https://github.com/juju/juju/blob/main/domain/errors.go#L19

Comment (wallyworld) [reply domain/application/errors/errors.go:13]: This is very deliberate. We absolutely must move away from "generic" not found errors and use proper, bespoke domain specific ones. We have specific errors already for machine not found, model not found etc.

**Review by SimonRichardson — APPROVED** [review]

File: domain/application/state/state.go [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+		logger:    logger,
+	}
+}
+
+// UpsertApplication creates or updates the specified application,
+// also adding any units if specified.
+// TODO - this just creates a minimal row for now.
+func (s *State) UpsertApplication(ctx context.Context, name string, units ...application.AddUnitParams) error {
```
Comment (manadart) [inline domain/application/state/state.go:43]: I don't know if the variadic args here are intended to reflect the API server multi-params, but we shouldn't do it where where we supply individual results in the returned payload object - we should have a transaction for each and return success/error in each case.

**Pull request merged**